File: examples/cpp/basic-bot.cpp
#include <dpp/dpp.h>
#include <iostream>
#include <fstream>
#include <string>
#include <map>

std::map<std::string, std::string> loadEnv() {
  std::map<std::string, std::string> env;
  std::ifstream file(".env");
  std::string line;
  while (std::getline(file, line)) {
    auto pos = line.find('=');
    if (pos != std::string::npos) {
      env[line.substr(0, pos)] = line.substr(pos + 1);
    }
  }
  return env;
}

int main() {
  auto env = loadEnv();
  std::string token = env["TOKEN"];
  std::string prefix = env["PREFIX"].empty() ? "!" : env["PREFIX"];

  dpp::cluster bot(token);

  bot.on_ready([&bot, &prefix](const dpp::ready_t& event) {
    if (dpp::run_once<0>()) {
      std::cout << "Logged in as " << bot.me.username << std::endl;

      dpp::slashcommand pingcmd("ping", "Replies with Pong!", bot.me.id);
      bot.global_command_create(pingcmd);
      std::cout << "Slash command registered" << std::endl;
    }
  });

  bot.on_slashcommand([](const dpp::slashcommand_t& event) {
    if (event.command.get_command_name() == "ping") {
      event.reply("Pong!");
    }
  });

  bot.on_message_create([&bot, &prefix](const dpp::message_create_t& event) {
    if (event.msg.author.is_bot()) return;
    std::string content = event.msg.content;

    if (content.find(prefix) != 0) return;
    std::string cmd = content.substr(prefix.length());

    if (cmd.rfind("ping", 0) == 0) {
      bot.message_create(event.msg.channel_id, "Pong!");
    }
  });

  bot.start(dpp::st_wait);
  return 0;
}

File: examples/lua/basic-bot.lua
local discordia = require('discordia')
local client = discordia.Client()

-- Load .env manually
local env = {}
local file = io.open('.env', 'r')
if file then
  for line in file:lines() do
    local key, val = line:match('^([^=]+)=(.*)$')
    if key then env[key] = val end
  end
  file:close()
end

local token = env.TOKEN
local prefix = env.PREFIX or '!'

client:on('ready', function()
  print('Logged in as ' .. client.user.username)

  -- Register slash command via REST API
  local https = require('ssl.https')
  local ltn12 = require('ltn12')
  local json = require('dkjson')

  local payload = json.encode({ name = 'ping', description = 'Replies with Pong!' })
  local url = 'https://discord.com/api/v10/applications/' .. client.application.id .. '/commands'

  local resp = {}
  https.request({
    url = url,
    method = 'POST',
    headers = {
      ['Content-Type'] = 'application/json',
      ['Authorization'] = 'Bot ' .. token,
      ['Content-Length'] = #payload,
    },
    source = ltn12.source.string(payload),
    sink = ltn12.sink.table(resp),
  })
  print('Slash command registered')
end)

client:on('interactionCreate', function(interaction)
  if interaction.type == 2 and interaction.data.name == 'ping' then
    interaction:reply({ content = 'Pong!' })
  end
end)

client:on('messageCreate', function(msg)
  if msg.author.bot then return end

  local content = msg.content
  if content:find(prefix, 1, true) ~= 1 then return end

  local args = {}
  for word in content:sub(#prefix + 1):gmatch('%S+') do
    table.insert(args, word)
  end

  if #args > 0 and args[1] == 'ping' then
    msg:reply('Pong!')
  end
end)

client:run('Bot ' .. token)
